File: src/android/app/src/main/java/com/mintreplica/lite/domain/usecase/AuthUseCases.kt
/*
 * Human Tasks:
 * 1. Configure proper API base URL in NetworkModule
 * 2. Set up ProGuard rules for Kotlin coroutines and Hilt
 * 3. Configure proper timeout and retry policies for API calls
 * 4. Set up proper SSL certificate pinning
 * 5. Configure biometric prompt strings in strings.xml
 */

// External library versions:
// - dagger.hilt.android: 2.44
// - kotlinx.coroutines: 1.6.4

package com.mintreplica.lite.domain.usecase

import androidx.fragment.app.FragmentActivity
import com.mintreplica.lite.data.repository.AuthRepository
import com.mintreplica.lite.utils.BiometricUtils
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Use case class that encapsulates authentication-related business logic
 * providing clean domain layer abstraction for authentication operations.
 *
 * Requirements addressed:
 * - 9.1.1 Authentication Methods: Implements secure user authentication with email/password and biometric options
 * - 9.1.3 Session Management: Handles authentication session with secure token storage
 * - 9.2.2 Data Security: Ensures secure handling of authentication credentials
 */
@Singleton
@AndroidEntryPoint
class AuthUseCases @Inject constructor(
    private val authRepository: AuthRepository
) {
    companion object {
        private const val MIN_PASSWORD_LENGTH = 8
        private const val EMAIL_PATTERN = "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+$"
        private const val MIN_NAME_LENGTH = 2
    }

    /**
     * Authenticates user with email and password credentials.
     * Validates input format and security requirements before authentication.
     */
    suspend fun loginWithCredentials(
        email: String,
        password: String
    ): Flow<Result<AuthToken>> = flow {
        try {
            // Validate email format
            if (!email.matches(EMAIL_PATTERN.toRegex())) {
                emit(Result.failure(Exception("Invalid email format")))
                return@flow
            }

            // Validate password requirements
            if (password.length < MIN_PASSWORD_LENGTH) {
                emit(Result.failure(Exception("Password must be at least $MIN_PASSWORD_LENGTH characters")))
                return@flow
            }

            // Perform login through repository
            authRepository.login(email, password).collect { result ->
                emit(result)
            }
        } catch (e: Exception) {
            emit(Result.failure(e))
        }
    }

    /**
     * Registers a new user account with secure password handling.
     * Validates all input requirements before registration.
     */
    suspend fun registerUser(
        email: String,
        password: String,
        name: String
    ): Flow<Result<AuthToken>> = flow {
        try {
            // Validate email format
            if (!email.matches(EMAIL_PATTERN.toRegex())) {
                emit(Result.failure(Exception("Invalid email format")))
                return@flow
            }

            // Validate password strength
            if (password.length < MIN_PASSWORD_LENGTH ||
                !password.any { it.isDigit() } ||
                !password.any { it.isUpperCase() } ||
                !password.any { it.isLowerCase() }
            ) {
                emit(Result.failure(Exception(
                    "Password must be at least $MIN_PASSWORD_LENGTH characters " +
                    "and contain at least one digit, one uppercase and one lowercase letter"
                )))
                return@flow
            }

            // Validate name requirements
            if (name.length < MIN_NAME_LENGTH || !name.trim().contains(" ")) {
                emit(Result.failure(Exception(
                    "Please provide both first and last name with minimum $MIN_NAME_LENGTH characters each"
                )))
                return@flow
            }

            // Perform registration through repository
            authRepository.register(email, password, name).collect { result ->
                emit(result)
            }
        } catch (e: Exception) {
            emit(Result.failure(e))
        }
    }

    /**
     * Authenticates user using FIDO2-compliant device biometrics.
     * Verifies biometric availability and configuration before authentication.
     */
    suspend fun loginWithBiometrics(
        activity: FragmentActivity
    ): Flow<Result<AuthToken>> = flow {
        try {
            // Check if biometric authentication is available
            if (!BiometricUtils.canAuthenticateWithBiometrics(activity)) {
                emit(Result.failure(Exception("Biometric authentication not available on this device")))
                return@flow
            }

            // Verify biometric is enabled for app
            if (!BiometricUtils.isBiometricEnabled(activity)) {
                emit(Result.failure(Exception("Biometric authentication not enabled for this app")))
                return@flow
            }

            // Perform biometric authentication through repository
            authRepository.authenticateWithBiometrics(activity).collect { result ->
                emit(result)
            }
        } catch (e: Exception) {
            emit(Result.failure(e))
        }
    }

    /**
     * Logs out the current user and cleans up session.
     * Ensures proper cleanup of authentication state and secure storage.
     */
    suspend fun logoutUser(): Flow<Result<Unit>> = flow {
        try {
            authRepository.logout()
            emit(Result.success(Unit))
        } catch (e: Exception) {
            emit(Result.failure(e))
        }
    }

    /**
     * Refreshes the authentication token before expiration.
     * Implements automatic token refresh to maintain session.
     */
    suspend fun refreshAuthToken(): Flow<Result<AuthToken>> = flow {
        try {
            // Check if user is authenticated
            if (!authRepository.isAuthenticated.value) {
                emit(Result.failure(Exception("User not authenticated")))
                return@flow
            }

            // Perform token refresh through repository
            authRepository.refreshToken().collect { result ->
                emit(result)
            }
        } catch (e: Exception) {
            emit(Result.failure(e))
        }
    }
}

File: src/ios/MintReplicaLite/Domain/UseCases/AuthUseCases.swift
//
// AuthUseCases.swift
// MintReplicaLite
//
// Implements authentication use cases with secure token management and biometric support

// MARK: - Human Tasks
/*
1. Verify proper keychain access group configuration in entitlements
2. Configure biometric authentication permissions in Info.plist
3. Set up proper SSL pinning for API security
4. Configure token refresh intervals in environment settings
5. Test biometric authentication flows across different device types
*/

import Foundation // iOS 14.0+
import Combine // iOS 14.0+

/// Implements authentication use cases with support for multiple authentication methods
/// and secure session management
/// Addresses requirements:
/// - Secure user authentication (1.2 Scope/Core Features)
/// - Authentication Methods (9.1.1 Authentication Methods)
/// - Session Management (9.1.3 Session Management)
@available(iOS 14.0, *)
final class AuthUseCases {
    
    // MARK: - Properties
    private let authRepository: AuthRepository
    private let biometricUtils: BiometricUtils
    private var cancellables = Set<AnyCancellable>()
    
    // MARK: - Initialization
    
    init(authRepository: AuthRepository, biometricUtils: BiometricUtils) {
        self.authRepository = authRepository
        self.biometricUtils = biometricUtils
    }
    
    // MARK: - Authentication Methods
    
    /// Authenticates user with email and password credentials
    /// - Parameters:
    ///   - email: User's email address
    ///   - password: User's password
    /// - Returns: Publisher emitting authenticated user or error
    func loginWithCredentials(email: String, password: String) -> AnyPublisher<User, Error> {
        // Validate input credentials
        guard isValidEmail(email) else {
            return Fail(error: ValidationError.invalidEmail).eraseToAnyPublisher()
        }
        
        guard isValidPassword(password) else {
            return Fail(error: ValidationError.invalidPassword).eraseToAnyPublisher()
        }
        
        // Attempt login through repository
        return authRepository.login(email: email, password: password)
            .handleEvents(receiveOutput: { [weak self] _ in
                // Setup biometric if enabled in preferences
                self?.setupBiometricIfEnabled(email: email, password: password)
            })
            .eraseToAnyPublisher()
    }
    
    /// Authenticates user using device biometric authentication (Face ID/Touch ID)
    /// - Returns: Publisher emitting authenticated user or error
    func loginWithBiometric() -> AnyPublisher<User, Error> {
        // Check biometric availability
        guard biometricUtils.isBiometricAvailable() else {
            return Fail(error: BiometricError.notAvailable).eraseToAnyPublisher()
        }
        
        // Attempt biometric authentication
        let result = biometricUtils.authenticateUser(reason: "Authenticate to access your account")
        
        switch result {
        case .success(true):
            // Perform login with stored credentials
            return authRepository.loginWithBiometric()
                .eraseToAnyPublisher()
            
        case .success(false):
            return Fail(error: BiometricError.canceled).eraseToAnyPublisher()
            
        case .failure(let error):
            return Fail(error: error).eraseToAnyPublisher()
        }
    }
    
    /// Registers a new user account with proper validation
    /// - Parameters:
    ///   - email: User's email address
    ///   - password: User's password
    ///   - firstName: User's first name
    ///   - lastName: User's last name
    /// - Returns: Publisher emitting created user or error
    func register(email: String, password: String, firstName: String, lastName: String) -> AnyPublisher<User, Error> {
        // Validate registration data
        guard isValidEmail(email) else {
            return Fail(error: ValidationError.invalidEmail).eraseToAnyPublisher()
        }
        
        guard isValidPassword(password) else {
            return Fail(error: ValidationError.invalidPassword).eraseToAnyPublisher()
        }
        
        guard isValidName(firstName) else {
            return Fail(error: ValidationError.invalidFirstName).eraseToAnyPublisher()
        }
        
        guard isValidName(lastName) else {
            return Fail(error: ValidationError.invalidLastName).eraseToAnyPublisher()
        }
        
        // Create user account through repository
        return authRepository.register(
            email: email,
            password: password,
            firstName: firstName,
            lastName: lastName
        )
        .eraseToAnyPublisher()
    }
    
    /// Logs out current user and clears authentication state securely
    /// - Returns: Publisher indicating logout success or error
    func logout() -> AnyPublisher<Void, Error> {
        return authRepository.logout()
            .handleEvents(receiveCompletion: { [weak self] completion in
                if case .finished = completion {
                    // Clear biometric state on successful logout
                    self?.biometricUtils.invalidateAuthentication()
                }
            })
            .eraseToAnyPublisher()
    }
    
    /// Refreshes the current authentication session with sliding expiration
    /// - Returns: Publisher indicating refresh success or error
    func refreshSession() -> AnyPublisher<Void, Error> {
        return authRepository.refreshToken()
            .eraseToAnyPublisher()
    }
    
    // MARK: - Private Methods
    
    private func isValidEmail(_ email: String) -> Bool {
        let emailRegex = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
        let emailPredicate = NSPredicate(format: "SELF MATCHES %@", emailRegex)
        return emailPredicate.evaluate(with: email)
    }
    
    private func isValidPassword(_ password: String) -> Bool {
        // Password requirements:
        // - At least 8 characters
        // - Contains at least one uppercase letter
        // - Contains at least one lowercase letter
        // - Contains at least one number
        // - Contains at least one special character
        let passwordRegex = "^(?=.*[A-Z])(?=.*[a-z])(?=.*\\d)(?=.*[$@$!%*?&])[A-Za-z\\d$@$!%*?&]{8,}"
        let passwordPredicate = NSPredicate(format: "SELF MATCHES %@", passwordRegex)
        return passwordPredicate.evaluate(with: password)
    }
    
    private func isValidName(_ name: String) -> Bool {
        // Name requirements:
        // - At least 2 characters
        // - Contains only letters and spaces
        // - No consecutive spaces
        let nameRegex = "^[A-Za-z]+(\\s{0,1}[A-Za-z]+)*$"
        let namePredicate = NSPredicate(format: "SELF MATCHES %@", nameRegex)
        return namePredicate.evaluate(with: name) && name.count >= 2
    }
    
    private func setupBiometricIfEnabled(email: String, password: String) {
        // Check if biometric authentication is enabled in user preferences
        guard UserDefaults.standard.bool(forKey: "biometric_enabled") else {
            return
        }
        
        // Verify biometric availability
        guard biometricUtils.isBiometricAvailable() else {
            return
        }
        
        // Store credentials for biometric authentication
        do {
            try authRepository.storeBiometricCredentials(email: email, password: password)
        } catch {
            print("Failed to store biometric credentials: \(error)")
        }
    }
}

// MARK: - Supporting Types

enum ValidationError: LocalizedError {
    case invalidEmail
    case invalidPassword
    case invalidFirstName
    case invalidLastName
    
    var errorDescription: String? {
        switch self {
        case .invalidEmail:
            return "Please enter a valid email address"
        case .invalidPassword:
            return "Password must be at least 8 characters and contain uppercase, lowercase, number and special character"
        case .invalidFirstName:
            return "Please enter a valid first name"
        case .invalidLastName:
            return "Please enter a valid last name"
        }
    }
}